config.txt
# PiZero model
[pi0]
# force_turbo = 1 in combination with over_voltage invalidates warranty!
# may improve compatibility, and needed for Steve's original kernels
force_turbo=1
arm_freq=1100
over_voltage=6
sdram_freq=450
over_voltage_sdram=2
boot_delay=1
arm_64bit=0
gpu_mem=0
#kernel_address=0x1f00000   # uncomment on PiZero, otherwise its used also for PiZero2W of 32bit boots... and fails
kernel=orig-build/kernel-Pi0.img

# this is for the PiZero 2W, merged with [p0]!
[pi02]
arm_freq=1400
over_voltage=4
sdram_freq=450
over_voltage_sdram=1
boot_delay=1
gpu_mem=0
hdmi_group=2
hdmi_mode=16
# select 32- or 64-bit mode
arm_64bit=0
# Pi 3 & Pi Zero 2W, 32bit
kernel=kernel8-32.img
# Pi 3 & Pi Zero 2W, 64bit
#kernel=kernel8.img

# P1 models
[pi1]
# force_turbo = 1 in combination with over_voltage invalidates warranty!
# may improve compatibility, and needed for Steve's original kernels
force_turbo=1
arm_freq=1100
over_voltage=6
sdram_freq=450
over_voltage_sdram=2
boot_delay=1
arm_64bit=0
gpu_mem=0
kernel_address=0x1f00000
kernel=orig-build/kernel-Pi1BPlus.img
#kernel=orig-build/kernel-Pi1BRev1.img
#kernel=orig-build/kernel-Pi1BRev2.img

[pi2] 
# untested
#arm_freq=1100
#sdram_freq=500
#over_voltage=6
kernel_address=0x1f00000
kernel=orig-build/kernel-Pi2.img

# all Pi3 variants
[pi3]
# force_turbo = 1 in combination with over_voltage invalidates warranty!
# may improve compatibility, and needed for Steve's original kernels
arm_freq=1400
force_turbo=1
over_voltage=4
sdram_freq=500
over_voltage_sdram=1
temp_soft_limit=70
temp_limit=80
boot_delay=1
gpu_mem=0
hdmi_group=2
hdmi_mode=16
# select 32- or 64-bit mode
arm_64bit=0     # set to 0 for original kernels!
# Pi 3 & Pi Zero 2W, 32bit
kernel=kernel8-32.img
# Pi 3 & Pi Zero 2W, 64bit
#kernel=kernel8.img
# Pi3 legacy 32bit kernel
#force_turbo = 1        # needed for original kernels, which are 32bit => also set "arm_64bit = 0" above!
#kernel_address=0x1f00000
#kernel=orig-build/kernel-Pi3.img

# all Pi4 variants
[pi4]
hdmi_force_hotplug=1   # Force HDMI output even if no monitor detected
hdmi_group=2           # Set display mode to DMT (monitor)
hdmi_mode=16           # Set resolution to 1024x768 @ 60Hz
max_framebuffers=2
# select 32- or 64-bit mode
arm_64bit=0
# Pi 4, 32bit
kernel=kernel7l.img
# Pi 4, 64bit
#kernel=kernel8-rpi4.img

[all]
#enable_uart=1
# Disable compensation for displays with overscan
disable_overscan=1


options.txt
// Sample options.txt configuration file for Pi1541
////////////////////////// HW Section

// If you are using the split line hardware option (ie Option B) then you need to specify this option
splitIECLines = 1
// If you are using some type of hardware that requires the lines to be inverted then specify these options here
//invertIECInputs = 1
// If you are using a 7407, set following option to 0
//invertIECOutputs = 0

// If you are using a LCD screen then specify it here
LCDName = ssd1306_128x64
//LCDName = ssd1306_128x32
//LCDName = sh1106_128x64
//LCDName = sh1107_128x128
// If you are using a LCD screen and you would like PageUp and PageDown keys to work with it (rather than the HDMI screen) then specify this option
//KeyboardBrowseLCDScreen = 1

// If you are using I2C LCD you can optionally change what pins it is connected to.
// (defaults to 0 for non-split lines (Option A) or 1 for split lines (Option B))
//i2cBusMaster = 0 //SDA - pin 27 SCL - pin 28
i2cBusMaster = 1 //SDA - pin 3 SCL - pin 5
i2cLcdAddress = 60	// I2C display address in decimal and shifted. 60 == 0x78, 61 == 0x7A
//i2cLcdFlip = 1 // Rotate i2c LCD screen 180 degrees
i2cLcdOnContrast = 127 // Allows you to adjust the contrast on your i2c LCD screen
i2cScan = 1 // scan i2c bus and display addresses on screen
//i2cLcdUseCBMChar = 1 // set it to 1 to use CBM font on LCD. Small but fun !

// If you have hardware with a peizo buzzer (the type without a generator) then you can use this option to hear the head step
// use SoundOnGPIO = 1 for GPIO
// use SoundOnGPIO = 0 for DMA Sound (=default)
// use SoundOnGPIO = -1 for disabling sound
SoundOnGPIO = 1
SoundOnGPIODuration = 1000 // Length of buzz in micro seconds
SoundOnGPIOFreq = 150 // Frequency of buzz in Hz

// You can remap the physical button functions
// numbers correspond to the standard board layout
//buttonEnter = 1
//buttonUp = 2
//buttonDown = 3
//buttonBack = 4
//buttonInsert = 5

//ROTARY:
//
// KY-040 Rotary Encoder Support
//
//    If you would like to use a KY-040 Rotary Encoder for browse menu up/down
//    and select, you can enable it here.  Connect as follows:
//
//       GPIO 22  -  Menu up       -  Encoder pin A (CLK)
//       GPIO 23  -  Menu down     -  Encoder pin B (DT)
//       GPIO 27  -  Enter/Select  -  Encoder pushbutton (SW)
//
//    ** Using an encoder is incompatible with the button remapping.  You must
//       use the default values of Enter=1, Up=2, Down=3, Back=4 and Insert=5.
//
//    ** This has only been tested using a Raspberry Pi 3.  Earlier models may
//       or may not work as expected!
//
//    Please see dmRotary.h for full implementation details.
//
//RotaryEncoderEnable = 1

////////////////////////// Emulation Section 

// If you would like to change the drive number then specify it here (8 is the default)
//deviceID = 9

// By default Pi1541 will search the root of the SD card for a ROM file and use this for ROM1.
// The first file found with either of the following names will be used; d1541.rom, dos1541, d1541II, Jiffy.bin
// You can override the default ROM1 by specifying it here
ROM1 = dos1541
//ROM1 = JiffyDOS_1541-II.bin
// You can specify additional ROMS here (up to 7)
// You can then use the function keys (F1-F7) on the keyboard to change which one you would like to use (only in browse mode and not in emulation mode)
// You can also use the buttons to swap ROMs by holding down button 1 and pressing one other buttons 2-5
// .LST files can also specify one of these ROMS that will automatically be selected when the LST file is loaded.
ROM2 = dos1541ii
//ROM3 = dos1541

// Reference you 1581 ROM here
ROM1581 = dos1581
//ROM1581 = JiffyDOS_1581.bin

// Enable RAMBoard emulation - 8k drive RAM expansion at 0x8000
//RAMBOard = 1

ChargenFont = chargen		// 8 bit font file

///////////////////////////////// Pi1541 Behavior
// The rate (in seconds) a long selection is scrolled
scrollHighlightRate = 0.07

// If using CBMFileBrowser then it is best to specify this option. When the computer resets the Pi will always revert back to the root folder ready to load CBMFileBrowser again.
OnResetChangeToStartingFolder = 1

// If you use FB64 (CBMFileBrowser) and want to use a fast loader cartridge (AR6, EFL, FC3) to load it then use this option to automatically mount it. 
AutoMountImage = fb.d64 // You MUST have a disk image in \1541 with this filename
// If you use FB64 (CBMFileBrowser) and want Pi1541 to send all file names as lower case.
LowercaseBrowseModeFilenames = 1

// If you are using a FB128 in 128 mode you can get FB128 to auto boot using this option
//AutoBootFB128 = 1

// If you are using a 128 you can auto boot anything using this option
// This over-rides AutoBootFB128
//128BootSectorName = bootsect.128

// If you would ever like to disable browse mode completely you can do so here
//DisableSD2IECCommands = 1

// change startup logo on oled - 1541ii or 1541classic
//LcdLogoName = 1541ii
LcdLogoName = 1541classic
//LcdLogoName = customfile.raw

// If you are using the Pi's composite video out then these options allow you to experiment with the display resolution
//ScreenWidth = 512
//ScreenHeight = 384

// You can create 320x200 PNG files with the same name as your disk images. With this option turned on they will be displayed on the Pi's screen.
DisplayPNGIcons = 1
// Display Tracks on HDMI
DisplayTrack = 0

// If you would like to specify what file will be loaded by LOAD"*" in browse mode then specify it here
//StarFileName = somefile

// Alt-N creates a new disk image.
// Names are formed automatically as autoname001.d64, autoname002.d64 etc...
// change the base of the filename here
//AutoBaseName = autoname
// By default Alt-N creates a new D64 disk image.
// If you rather it create a G64 then use this option.
//NewDiskType = g64

QuickBoot = 0		// faster startup
ShowOptions = 0	// display some options on startup screen 
IgnoreReset = 0

// This option will display the temperature of the Pi's CPU.
// It should be about 52C anything above 75C is bad and there is something wrong with your hardware
DisplayTemperature = 0  // may impact compatibility / timing of emulation
DisableHDMI = 0         // same as headLess, compatibility with recent Pi1541 (>V1.24), better set to 1 on Pi0s and small mem Pis (Pi3A+)
// This option displays the IEC bus activity on the bottom of the Pi's screen
HDMIGraphIEC = 1        // only if HDMI is enabled, don't use on on PiZero2
HDMIDisplayIECActivity = 1    // only if HDMI is enabled

/////////////////////////////// pottendo Pi1541 Options
// this turns on/off HDMI output
// headLess = 1 // replaced by DisableHDMI

// Network configuration, WiFi needs a proper wpa_supplicant.conf configured for you SSID
TZ = 2.0    // 2.0=CEST is default, use -5.0 for ET (e.g. New York/US), use 5.5 for New Delhi/IN
netWifi = 0
netEthernet = 0

// webserver static memory profile in kB, use smaller value (4000) for Pis < 1GB memory.
maxContentSize = 1000     // == 1MB (min 256, max 4000)
maxMultipartSize = 20000  // == 20MB for directory upload (min 1000, max 20000)

// Static network config, to avoid slow DHCP - not much sanity is done, so write properly
useDHCP = 1 // get network config automatically, else set to '0' and define static network config below
//IPAdress = 192.168.1.31
//NetMask = 192.168.1.0
//DefaultGateway = 192.168.1.1
//DNSServer = 192.168.1.1


src/options.cpp
// Pi1541 - A Commodore 1541 disk drive emulator
// Copyright(C) 2018 Stephen White
//
// This file is part of Pi1541.
// 
// Pi1541 is free software : you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// Pi1541 is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with Pi1541. If not, see <http://www.gnu.org/licenses/>.

#include "options.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h>
#include <ctype.h>

#define INVALID_VALUE	((int)0xDEADBEEF)

char* TextParser::GetToken(bool includeSpace)
{
	bool isSpace;

	do
	{
	} while (ParseComment());

	while (*data != '\0')
	{
		isSpace = isspace(*data);

		if (!isSpace || (includeSpace && isSpace))
			break;

		data++;
	}

	if (*data == '\0')
		return 0;

	char* pToken = data;

	while (*data != '\0')
	{
		isSpace = isspace(*data);

		if ((!includeSpace && isSpace) || (*data == '\n') || (*data == '\r'))
		{
			*data++ = '\0';
			break;
		}
		data++;
	}

	return pToken;
}

void TextParser::SkipWhiteSpace()
{
	while (*data != '\0')
	{
		if (!isspace(*data))
			break;

		data++;
	}
}

bool TextParser::ParseComment()
{
	SkipWhiteSpace();

	if (*data == '\0')
		return 0;

	if (data[0] != '/')
		return false;

	if (data[1] == '/')
	{
		// One line comment
		data += 2;

		while (*data)
		{
			if (*data == '\n')
				break;

			data++;
		}
		SkipWhiteSpace();
		return true;
	}
	else if (data[1] == '*')
	{
		// Multiline comment
		data += 2;

		while (*data)
		{
			if (*data++ == '*' && *data && *data == '/')
			{
				data++;
				break;
			}
		}
		SkipWhiteSpace();
		return true;
	}
	return false;
}


Options::Options(void)
	: TextParser()
	, deviceID(8)
	, onResetChangeToStartingFolder(0)
	, extraRAM(0)
	, RAMBOard(0)
	, disableSD2IECCommands(0)
	, disableHDMI(0)
	, supportUARTInput(0)
	, hdmiGraphIEC(0)
	, hdmiDisplayIECActivity(0)
	, displayTracks(0)
	, quickBoot(0)
	, logoDisplayDelay(0)
	, showOptions(0)
	, displayPNGIcons(0)
	, soundOnGPIO(0)
	, soundOnGPIODuration(1000)
	, soundOnGPIOFreq(1200)
	, invertIECInputs(0)
	, invertIECOutputs(1)
	, splitIECLines(0)
	, ignoreReset(0)
	, autoBootFB128(0)
	, displayTemperature(0)
	, lowercaseBrowseModeFilenames(0)
	, screenWidth(1024)
	, screenHeight(768)
	, i2cBusMaster(1)
	, i2cLcdAddress(0x3C)
	, i2cScan(0)
	, i2cLcdFlip(0)
	, i2cLcdOnContrast(127)
	, i2cLcdUseCBMChar(0)
	, i2cLcdModel(LCD_UNKNOWN)
	, scrollHighlightRate(0.125f)
	, keyboardBrowseLCDScreen(0)
        , buttonEnter(1)
        , buttonUp(2)
        , buttonDown(3)
        , buttonBack(4)
        , buttonInsert(5)
	, rotaryEncoderEnable(0) //ROTARY:
	, rotaryEncoderInvert(0) //ROTARY:
	, headLess(0)
#if defined(__CIRCLE__)	
	, netWifi(0)
	, netEthernet(0)
	, maxContentSize(1000)
	, maxMultipartSize(20000)
	, noHealthMonitor(1)
	, useDHCP(1)
	, TZ(2.0)
#endif	
{
	autoMountImageName[0] = 0;
	strcpy(ROMFontName, "chargen");
	strcpy(LcdLogoName, "1541ii");
	strcpy(autoBaseName, "autoname");
	C128BootSectorName[0] = 0;
	starFileName[0] = 0;
	ROMName[0] = 0;
	ROMNameSlot2[0] = 0;
	ROMNameSlot3[0] = 0;
	ROMNameSlot4[0] = 0;
	ROMNameSlot5[0] = 0;
	ROMNameSlot6[0] = 0;
	ROMNameSlot7[0] = 0;
	ROMNameSlot8[0] = 0;
	ROMName1581[0] = 0;
	ROMName1571[0] = 0;
	newDiskType[0] = 0;
	i2cLcdModelName = "LCD_UNKNOWN";
}

#define ELSE_CHECK_DECIMAL_OPTION(Name) \
	else if (strcasecmp(pOption, #Name) == 0) \
	{ \
		int nValue = 0; \
		if ((nValue = GetDecimal(pValue)) != INVALID_VALUE) \
			Name = nValue; \
	}
#define ELSE_CHECK_FLOAT_OPTION(Name) \
	else if (strcasecmp(pOption, #Name) == 0) \
	{ \
		float value = 0; \
		if ((value = GetFloat(pValue)) != INVALID_VALUE) \
			Name = value; \
	}

void Options::Process(char* buffer)
{
	SetData(buffer);

	char* pOption;
	while ((pOption = GetToken()) != 0)
	{
		/*char* equals = */GetToken();
		char* pValue = GetToken();

		if ((strcasecmp(pOption, "Font") == 0) || (strcasecmp(pOption, "ChargenFont") == 0))
		{
			strncpy(ROMFontName, pValue, 255);
		}
		else if ((strcasecmp(pOption, "AutoMountImage") == 0))
		{
			strncpy(autoMountImageName, pValue, 255);
		}
		ELSE_CHECK_DECIMAL_OPTION(deviceID)
		ELSE_CHECK_DECIMAL_OPTION(onResetChangeToStartingFolder)
		ELSE_CHECK_DECIMAL_OPTION(extraRAM)
		ELSE_CHECK_DECIMAL_OPTION(RAMBOard)
		ELSE_CHECK_DECIMAL_OPTION(disableSD2IECCommands)
		ELSE_CHECK_DECIMAL_OPTION(disableHDMI)
		ELSE_CHECK_DECIMAL_OPTION(supportUARTInput)
		ELSE_CHECK_DECIMAL_OPTION(hdmiGraphIEC)
		ELSE_CHECK_DECIMAL_OPTION(hdmiDisplayIECActivity)
		ELSE_CHECK_DECIMAL_OPTION(displayTracks)
		ELSE_CHECK_DECIMAL_OPTION(quickBoot)
		ELSE_CHECK_DECIMAL_OPTION(logoDisplayDelay)
		ELSE_CHECK_DECIMAL_OPTION(showOptions)
		ELSE_CHECK_DECIMAL_OPTION(displayPNGIcons)
		ELSE_CHECK_DECIMAL_OPTION(soundOnGPIO)
		ELSE_CHECK_DECIMAL_OPTION(soundOnGPIODuration)
		ELSE_CHECK_DECIMAL_OPTION(soundOnGPIOFreq)
		ELSE_CHECK_DECIMAL_OPTION(invertIECInputs)
		ELSE_CHECK_DECIMAL_OPTION(invertIECOutputs)
		ELSE_CHECK_DECIMAL_OPTION(splitIECLines)
		ELSE_CHECK_DECIMAL_OPTION(ignoreReset)
		ELSE_CHECK_DECIMAL_OPTION(lowercaseBrowseModeFilenames)
		ELSE_CHECK_DECIMAL_OPTION(autoBootFB128)
		ELSE_CHECK_DECIMAL_OPTION(displayTemperature)
		ELSE_CHECK_DECIMAL_OPTION(screenWidth)
		ELSE_CHECK_DECIMAL_OPTION(screenHeight)
		ELSE_CHECK_DECIMAL_OPTION(i2cBusMaster)
		ELSE_CHECK_DECIMAL_OPTION(i2cLcdAddress)
		ELSE_CHECK_DECIMAL_OPTION(i2cScan)
		ELSE_CHECK_DECIMAL_OPTION(i2cLcdFlip)
		ELSE_CHECK_DECIMAL_OPTION(i2cLcdOnContrast)
		ELSE_CHECK_DECIMAL_OPTION(i2cLcdDimContrast)
		ELSE_CHECK_DECIMAL_OPTION(i2cLcdDimTime)
		ELSE_CHECK_DECIMAL_OPTION(i2cLcdUseCBMChar)
		ELSE_CHECK_FLOAT_OPTION(scrollHighlightRate)
		ELSE_CHECK_DECIMAL_OPTION(keyboardBrowseLCDScreen)
		ELSE_CHECK_DECIMAL_OPTION(buttonEnter)
		ELSE_CHECK_DECIMAL_OPTION(buttonUp)
		ELSE_CHECK_DECIMAL_OPTION(buttonDown)
		ELSE_CHECK_DECIMAL_OPTION(buttonBack)
		ELSE_CHECK_DECIMAL_OPTION(buttonInsert)
		ELSE_CHECK_DECIMAL_OPTION(rotaryEncoderEnable) //ROTARY:
		ELSE_CHECK_DECIMAL_OPTION(rotaryEncoderInvert) //ROTARY:
		ELSE_CHECK_DECIMAL_OPTION(headLess)
#if defined(__CIRCLE__)
		ELSE_CHECK_DECIMAL_OPTION(netWifi)
		ELSE_CHECK_DECIMAL_OPTION(netEthernet)
		ELSE_CHECK_DECIMAL_OPTION(maxContentSize)
		ELSE_CHECK_DECIMAL_OPTION(maxMultipartSize)
		ELSE_CHECK_DECIMAL_OPTION(noHealthMonitor)
		ELSE_CHECK_DECIMAL_OPTION(useDHCP)
		ELSE_CHECK_FLOAT_OPTION(TZ)
#endif		
		else if ((strcasecmp(pOption, "AutoBaseName") == 0))
		{
			strncpy(autoBaseName, pValue, 255);
		}
		else if ((strcasecmp(pOption, "128BootSectorName") == 0))
		{
			strncpy(C128BootSectorName, pValue, 255);
		}
		else if ((strcasecmp(pOption, "StarFileName") == 0))
		{
			strncpy(starFileName, pValue, 255);
		}
		else if ((strcasecmp(pOption, "LCDLogoName") == 0))
		{
			strncpy(LcdLogoName, pValue, 255);
		}
		else if ((strcasecmp(pOption, "LCDName") == 0))
		{
			strncpy(LCDName, pValue, 255);
			//DEBUG_LOG("%s: lcd selected: '%s'", __FUNCTION__, LCDName);
			if (strcasecmp(pValue, "ssd1306_128x64") == 0)
			{
				i2cLcdModel = LCD_1306_128x64;
				i2cLcdModelName = "LCD_1306_128x64";
			}
			else if (strcasecmp(pValue, "ssd1306_128x32") == 0)
			{
				i2cLcdModel = LCD_1306_128x32;
				i2cLcdModelName = "LCD_1306_128x32";
			}
			else if (strcasecmp(pValue, "sh1106_128x64") == 0)
			{
				i2cLcdModel = LCD_1106_128x64;
				i2cLcdModelName = "LCD_1106_128x64";
			}
			else if (strcasecmp(pValue, "sh1107_128x128") == 0)
			{
				i2cLcdModel = LCD_1107_128x128;
				i2cLcdModelName = "LCD_1107_128x128";
			}
			else 
			{
				i2cLcdModel = LCD_UNKNOWN;
				i2cLcdModelName = "LCD_UNKNOWN";
			}
		}
		else if ((strcasecmp(pOption, "ROM1581") == 0))
		{
			strncpy(ROMName1581, pValue, 255);
		}
		else if ((strcasecmp(pOption, "ROM1571") == 0))
		{
			strncpy(ROMName1571, pValue, 255);
		}
		else if ((strcasecmp(pOption, "ROM") == 0) || (strcasecmp(pOption, "ROM1") == 0))
		{
			strncpy(ROMName, pValue, 255);
		}
		else if (strcasecmp(pOption, "ROM2") == 0)
		{
			strncpy(ROMNameSlot2, pValue, 255);
		}
		else if (strcasecmp(pOption, "ROM3") == 0)
		{
			strncpy(ROMNameSlot3, pValue, 255);
		}
		else if (strcasecmp(pOption, "ROM4") == 0)
		{
			strncpy(ROMNameSlot4, pValue, 255);
		}
		else if (strcasecmp(pOption, "ROM5") == 0)
		{
			strncpy(ROMNameSlot5, pValue, 255);
		}
		else if (strcasecmp(pOption, "ROM6") == 0)
		{
			strncpy(ROMNameSlot6, pValue, 255);
		}
		else if (strcasecmp(pOption, "ROM7") == 0)
		{
			strncpy(ROMNameSlot7, pValue, 255);
		}
		else if (strcasecmp(pOption, "ROM8") == 0)
		{
			strncpy(ROMNameSlot8, pValue, 255);
		}
		else if ((strcasecmp(pOption, "NewDiskType") == 0))
		{
			strncpy(newDiskType, pValue, 31);
		}
#if defined(__CIRCLE__)
		else if (strcasecmp(pOption, "IPAdress") == 0)
		{
			setIP(pValue);
		}
		else if (strcasecmp(pOption, "NetMask") == 0)
		{
			setNM(pValue);
		}
		else if (strcasecmp(pOption, "DefaultGateway") == 0)
		{
			setGW(pValue);
		}
		else if (strcasecmp(pOption, "DNSServer") == 0)
		{
			setDNS(pValue);
		}
#endif
	}

	if (!SplitIECLines())
	{
		invertIECInputs = false;
		// If using non split lines then only the 1st bus master can be used (as ATN is using the 2nd)
		i2cBusMaster = 0;
	}
}

int Options::GetDecimal(char* pString)
{
	if (pString == 0 || *pString == '\0')
		return 0;

	return strtol(pString, NULL, 0);
}

float Options::GetFloat(char* pString)
{
	if (pString == 0 || *pString == '\0')
		return 0;

	return atof(pString);
}

const char* Options::GetRomName(int index) const
{
	switch (index)
	{
	case 1:
		return ROMNameSlot2;
	case 2:
		return ROMNameSlot3;
	case 3:
		return ROMNameSlot4;
	case 4:
		return ROMNameSlot5;
	case 5:
		return ROMNameSlot6;
	case 6:
		return ROMNameSlot7;
	case 7:
		return ROMNameSlot8;
	}
	return ROMName;
}

const char* Options::GetRomName1581() const
{
	if (ROMName1581[0] == 0)
		return "1581-rom.318045-02.bin";
	else
		return ROMName1581;
}

const char* Options::GetRomName1571() const
{
	if (ROMName1571[0] == 0)
		return "dos1571-310654-05.bin";
	else
		return ROMName1571;
}

DiskImage::DiskType Options::GetNewDiskType() const
{
	if (strcasecmp(newDiskType, "g64") == 0)
		return DiskImage::G64;

	return DiskImage::D64;
}

